org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/EclipseResourceFileSystemAccess2.java

```diff
@@ -409,7 +418,55 @@ protected boolean hasContentsChanged(IFile file, InputStream newContent) {
 		}
 		return contentChanged;
 	}
-
+	
+	/**
+	 * @since 2.30
+	 */
+	protected boolean hasContentsChangedNew(InputStream oldContent, InputStream newStream) {
+		// Since we don't know if the code may run in multiple threads, use thread local caches
+		byte [] oldBytes = OLD_BYTES.get();
+        byte [] newBytes = NEW_BYTES.get();
+        boolean contentChanged;
+        int newRead;
+        int oldRead;
+        try (InputStream oldStream = oldContent){
+            do {
+                oldRead = read(oldStream, oldBytes, 0, DEFAULT_BUFFER_SIZE);
+                newRead = read(newStream, newBytes, 0, DEFAULT_BUFFER_SIZE);
+                if(newRead != oldRead) {
+                    contentChanged = true;
+                    break;
+                }
+                contentChanged = !Arrays.equals(oldBytes, 0, oldRead, newBytes, 0, newRead);
+                if(newRead < DEFAULT_BUFFER_SIZE) {
+                    // EOF
+                    break;
+                }
+            } while (!contentChanged);
+
+        } catch (IOException e) {
+            contentChanged = true;
+        }
+        return contentChanged;
+	}
+
+	/**
+	 * @since 2.30
+	 */
+    protected int read(InputStream in, byte[] b, int off, int len) throws IOException {
```

Comment:
Since this is essentially `com.google.common.io.ByteStreams.read(InputStream, byte[], int, int)`, maybe use that one or add a comment why it wasn't used.